docs/development/design/system-packages.rst

```diff
@@ -0,0 +1,100 @@
+Allow Installation of System Packages
+=====================================
+
+Currently we don't allow executing arbitrary commands in the build process.
+The more common use case is to install extra dependencies.
+
+.. contents::
+   :local:
+   :depth: 3
+
+Current status
+--------------
+
+There is a workaround when using Sphinx to run arbitrary commands,
+this is executing the commands inside the ``conf.py`` file.
+There isn't a workaround for MkDocs, but this problem is more common in Sphinx,
+since users need to install some extra dependencies in order to use autodoc or build Jupyter Notebooks.
+
+However, installation of some dependencies require root access,
+or are easier to install using ``apt``.
+Most of the CI services allow to use ``apt`` or execute any command with ``sudo``,
+so users are more familiar with that workflow.
+
+Some users use Conda instead of pip to install dependencies in order to avoid these problems,
+but not all pip users are familiar with Conda, or want to migrate to Conda just to use Read the Docs.
+
+Security concerns
+-----------------
+
+Builds are run in a Docker container,
+but the app controlling that container lives in the same server.
+Allowing to execute arbitrary commands with super user privileges may introduce some security issues.
+
+Exposing ``apt install``
+------------------------
+
+For the previous reasons we won't allow to execute arbitrary commands with root (yet),
+but instead allow only to install extra packages using ``apt``.
+
+We would expose this through the config file.
+Users will provide a list of packages to install, and under the hook we would run:
+
+- ``apt update -y``
+- ``apt install -y {packages}``
+
+These commands will be run before the Python setup step and after the clone step.
+
+Using ``docker exec``
+---------------------
+
+Currently we user ``docker exec`` to execute commands in a running container.
+This command also allows to pass a user which is used to run the commands (#8058_).
+We can run the ``apt`` commands in our current containers using a super user momentarily.
+
+.. _#8058: https://github.com/readthedocs/readthedocs.org/pull/8058
+
+Config file
+-----------
+
+The config file can add an additional mapping ``build.packages`` to a list of packages to install.
+
+.. code-block:: yaml
+
+   version: 2
+
+   build:
+     packages:
+        - cmatrix
+        - mysql-server
+
+.. note::
+
+   We can rename the key to be ``build.apt_packages`` or ``build.extra_packages``
+   so users aren't confused with Python packages or maybe having that option under the ``build`` key is enough.
```

Comment:
`system_packages` is already used in the python options https://docs.readthedocs.io/en/stable/config-file/v2.html#python-system-packages

Reply:
Is that a problem? I think it's fine to have `python.system_packages` and `build.system_packages`.

Reply:
Reading your code, https://github.com/readthedocs/readthedocs.org/pull/8065/files#diff-99dfd740f4f146e66e1c3709bebdc3db357a8c6defbf3e592434be4308f5d697R1156, I realized that another possible name could be `build.system_dependencies`

Reply:
> Is that a problem? I think it's fine to have `python.system_packages` and `build.system_packages`.

Potentially confusing I'd say. Moreover, the former is a boolean and the latter is a list.

In retrospect, the former could have been called `inherit_system_site_packages`, but it's both too long and too late :)

Although @davidfischer suggested on https://github.com/readthedocs/readthedocs.org/pull/7566#discussion_r596207338 using a name that doesn't tie us to apt-based distros, I agree with @ericholscher that this is not bulletproof at all. Imagine that dependency X is called `libx` in apt and `libx1` in yum: changing the base image/package manager would break builds declaring `libx` as dependency, or produce difficult to debug errors.

Reply:
Yea, I still think having it be explicit is good. Currently we only support `apt`, but maybe in the future we support other package managers, etc? I think making the naming as explicit as possible is good. We _could_ name it `system_packages`, but I think in the short-term it's confusing (we only support apt), and in the long-term I'm not convinced that `system_packages` is good if we support multiple architectures (eg. apt & rpm) or (apt & nuget), given that they map to different namespaces. Is having `apt_packages`, `rpm_packages`, etc. an issue?

Reply:
We only support apt right now but if we ever move to support custom docker containers, those custom containers may not support apt. I still vote for `system_packages` but I don't care enough to really fight it.

Reply:
Sure, but having ambiguous `system_packages` that aren't mapped to a specific namespace feels even more broken, doesn't it?

Reply:
I don't have a strong opinion on this, but mostly because I hope we could allow to execute any commands in the future instead of relying on this option p:

Reply:
> Sure, but having ambiguous system_packages that aren't mapped to a specific namespace feels even more broken, doesn't it?

Maybe `system_dependencies` is more precise? Apt feels very much like an implementation detail.

Maybe Santos is right that long term we should get toward custom commands and maybe custom containers which seems closer to what the CI companies are doing.

Reply:
Apt isn't an implementation detail. It's very user facing, because it's the namespace that the key represents. You can't just put "docker" or "git" in that field and have it work.